### DISCUSSION TITLE
SSDP for device management

### DISCUSSION THREAD

**Discussion by EliRibble** [arendst/Tasmota#19287] [Ideas]
I've recently been working on switching all 60 light switches in my home to a Tasmota-based smart switch. Managing this many switches is laborious one-by-one. Is there a method for programmatically discovering all devices on a given network once I've got them on wifi? For example, [SSDP](https://en.wikipedia.org/wiki/Simple_Service_Discovery_Protocol) (part of uPnP) which uses UDP packets multicast addresses to search for devices providing specific services.

Ideally I'd like something pretty simple where I can send a packet like:

```
M-SEARCH * HTTP/1.1
Host: 239.255.255.250:1900
MAN: "ssdp:discover"
MX: 3
ST: urn:tasmota
```

It could be interesting to separate out devices by capability, so ST could end with ":switch", ":sensor", ":thermostat", etc. Response from a device would be something like:

```
HTTP/1.1 200 OK
CACHE-CONTROL: max-age=3600
ST: urn:tasmota:device
USN: uuid:cf367a59-d637-4181-877f-e1963a6f32be::urn:tasmota:device
EXT:
SERVER: Tasmota 13.0.1
LOCATION: http://192.168.1.105:80/cs
```

If this doesn't exist and the project owners are interested I'd be willing to take a crack at implementing it.

**Comment by barbudor** [comment]
Are you using MQTT?
If so tasmota is already providing a discovery protocol
Look at topics under 'tasmota/'

But 1st question is how are you provisionning the devices on your network?

**Reply by EliRibble** [reply]
I'm provisioning them with my phone via the standalone `tasmota_(id)` network, once for each device. It sucks, but I don't know any other way.

I was hoping to use the discovery protocol to do the heavy lifting for getting the devices on MQTT and other configuration steps.

**Reply by sfromis** [reply]
To me, it is the other way around. I make sure to config for MQTT right after provisioning, and then MQTT makes all the rest easy to do, as Tasmota publishes retained payloads providing for easy discovery just by subscribing to  `tasmota/discovery/+/config`. That way, I get all devices right away.

If I wanted to scan my network for Tasmota devices, I could do a http sweep looking for headers with `server: Tasmota .....`, but haven't needed that.

**Reply by sfromis** [reply]
An easier way of provisioning would be if you create a custom binary with your Wifi/MQTT details "baked in". Then devices can appear on your network right after flashing.

To be clear, my normal way of provisioning is via serial, as the large majority of devices is flashed that way, and when serial is already connected, it is easy to shoot a few commands.

**Reply by EliRibble** [reply]
Fair point, I'm buying devices that are pre-flashed, so I don't have this step, but it makes sense to hook at this step if you do.

**Reply by EliRibble** [reply]
Ah, the "Tasmota" server header. That's something I missed. That's pretty good!
